aws-s3-transfer-manager/tests/upload_checksum_test.rs

```diff
@@ -0,0 +1,829 @@
+
+async fn test_put_object(
+    user_checksum_strategy: Option<ChecksumStrategy>,
+    body: Bytes,
+) -> UploadOutput {
+    upload_helper(
+        user_checksum_strategy,
+        vec![body],
+        false,
+        aws_sdk_s3::config::RequestChecksumCalculation::WhenSupported,
+    )
+    .await
+}
+
+async fn upload_helper(
+    user_checksum_strategy: Option<ChecksumStrategy>,
+    parts: Vec<Bytes>,
+    send_as_multipart: bool,
+    sdk_checksum_calculation: aws_sdk_s3::config::RequestChecksumCalculation,
+) -> UploadOutput {
+    let (_guard, _rx) = capture_test_logs();
+
+    // This is the ChecksumStrategy we expect the Transfer Manager to use while sending requests
+    let request_checksum_strategy = user_checksum_strategy.clone().or(
+        // If user didn't set a strategy, Transfer Manager should fall back to default strategy,
+        // unless user also disabled checksums via SDK's config.
+        if sdk_checksum_calculation == aws_sdk_s3::config::RequestChecksumCalculation::WhenSupported
+        {
+            Some(ChecksumStrategy::default())
+        } else {
+            None
+        },
+    );
+
+    // Determine which kind of checksums will be in the response.
+    // As of 2025, S3 always provides checksums in response, even if request had none.
+    let (response_checksum_algorithm, response_checksum_type) = match &request_checksum_strategy {
+        None => (ChecksumAlgorithm::Crc64Nvme, ChecksumType::FullObject),
+        Some(request_checksum_strategy) => (
+            request_checksum_strategy.algorithm.clone(),
+            request_checksum_strategy.type_if_multipart.clone(),
+        ),
+    };
+
+    // Create mock aws_sdk_s3::Client
+    let s3_client = if send_as_multipart {
+        mock_s3_client_for_multipart_upload(
+            request_checksum_strategy.clone(),
+            response_checksum_algorithm.clone(),
+            response_checksum_type.clone(),
+            &parts,
+        )
+    } else {
+        assert_eq!(parts.len(), 1);
+        mock_s3_client_for_put_object(
+            request_checksum_strategy.clone(),
+            response_checksum_algorithm.clone(),
+            &parts[0],
+        )
+    };
+
+    // Adjust client's SDK config to use specific RequestChecksumCalculation
+    let s3_client = aws_sdk_s3::Client::from_conf(
+        s3_client
+            .config()
+            .to_builder()
+            .request_checksum_calculation(sdk_checksum_calculation.clone())
+            .build(),
+    );
+
+    let body_stream: InputStream = if send_as_multipart {
+        InputStream::from_part_stream(TestStream::new(parts.clone()))
+    } else {
+        parts[0].clone().into()
+    };
+
+    let tm_config = aws_s3_transfer_manager::Config::builder()
+        .client(s3_client)
+        .part_size(PartSize::Target(PART_SIZE as u64))
+        .multipart_threshold(PartSize::Target(PART_SIZE as u64))
+        .concurrency(ConcurrencySetting::Explicit(1)) // guarantee parts sent in order
```

Comment:
This is good/safe for unit tests. Do we need/want to consider one or more tests against real S3?

Reply:
That's an excellent idea, but I don't want to spend a week adding the repo's very first integration test. Let's tackle this along with:
- https://github.com/awslabs/aws-s3-transfer-manager-rs/issues/25